src/Vendr.PaymentProviders.Worldpay/WorldpayPaymentProvider.cs

```diff
@@ -1,38 +1,139 @@
+                settings.OrderPropertyBillingLastName.MustNotBeNull("settings.OrderPropertyBillingLastName");
+
+                var firstname = order.Properties[settings.OrderPropertyBillingFirstName];
+                var surname = order.Properties[settings.OrderPropertyBillingLastName];
+                var address1 = order.Properties[settings.OrderPropertyBillingAddress1] ?? string.Empty;
+                var city = order.Properties[settings.OrderPropertyBillingCity] ?? string.Empty;
+                var postcode = order.Properties[settings.OrderPropertyBillingPostcode] ?? string.Empty;
+                var billingCountry = Vendr.Services.CountryService.GetCountry(order.PaymentInfo.CountryId.Value);
+                var billingCountryCode = billingCountry.Code.ToUpperInvariant();
+                var amount = order.TotalPrice.Value.WithTax.ToString("0.00", CultureInfo.InvariantCulture);
+                var currency = Vendr.Services.CurrencyService.GetCurrency(order.CurrencyId);
+                var currencyCode = currency.Code.ToUpperInvariant();
+
+                // Ensure billing country has valid ISO 3166 code
+                var iso3166Countries = Vendr.Services.CountryService.GetIso3166CountryRegions();
+                if (iso3166Countries.All(x => x.Code != billingCountryCode))
+                {
+                    throw new Exception("Country must be a valid ISO 3166 billing country code: " + billingCountry.Name);
+                }
+
+                // Ensure currency has valid ISO 4217 code
+                if (!Iso4217.CurrencyCodes.ContainsKey(currencyCode))
+                {
+                    throw new Exception("Currency must be a valid ISO 4217 currency code: " + currency.Name);
+                }
+
+                var orderDetails = new Dictionary<string, string>
+                {
+                    { "instId", settings.InstallId },
+                    { "testMode", settings.TestModeNumber },
+                    { "authMode", settings.AuthMode },
+                    { "cartId", order.CartNumber },
+                    { "amount", amount },
+                    { "currency", currencyCode },
+                    { "MC_cancelurl", cancelUrl },
+                    { "MC_returnurl", continueUrl },
+                    { "MC_callbackurl", callbackUrl },
+                    { "name", firstname + " " + surname },
+                    { "email", order.CustomerInfo.Email },
+                    { "address1", address1 },
+                    { "town", city },
+                    { "postcode", postcode },
+                    { "country", billingCountryCode }
+                };
+
+                if (!string.IsNullOrEmpty(settings.Md5Secret))
+                {
+                    var orderSignature = Md5Helper.CreateMd5(settings.Md5Secret + ":" + amount + ":" + currencyCode + ":" + settings.InstallId + ":" + order.CartNumber);
```

Comment:
I would probably use `order.OrdrNumber` instead of `order.CartNumber`. Vendr is slightly different to TC in that it generates an order number prior to going to the payment gateway.